R/Graph.R

```diff
@@ -206,3 +214,66 @@ graph_plot = function(root) {
 
   plot(g, layout = layout$layout)
 }
+
+#' graph_map_topo
+#'
+#' @param root root node of the graph.
+#' @param fnc function to apply
+#' @param simplify should the result be simplified using simplify2array.
+#' Default TRUE.
+#' @param add_layer if true add information about the layer as a attribute.
+#'
+#' @return
+#'
+#' List (possibly simplified to vector) with the output
+#' of function `fnc`  applied to all the nodes of the graph
+#' in topological order. Note that only the output is in
+#' topological order. The function `fnc` is not applied in that order,
+#' contrary, the algorithm uses a `depth-first search`,
+#' so it starts applying the function from the last element.
+#'
+#' @noRd
+#'
+graph_map_topo = function(root, fnc = function(x) x$id, simplify = TRUE, add_layer = FALSE) {
+
+  state = new.env()
+  state$permanent = c()
+  state$temporary = c()
+  state$list = list()
+  state$depth = numeric()
+
+  visit = function(node, state, depth = 1) {
+    if(node$id %in% state$permanent) {
+      state$depth[node$id] = pmax(depth, state$depth[node$id], na.rm = TRUE)
+      return()
+    }
+    if(node$id %in% state$temporary) stop("Not a DAG")
+    state$temporary = c(node$id, state$temporary) # mark temporarily
+
+    if(length(node$next_nodes) > 0) {
```

Comment:
Do we want a helper function here, so we consistently check this across multiple functions that require this.

Reply:
I added an AB method `has_next`.